ding/reward_model/count_based_model.py

```diff
@@ -0,0 +1,354 @@
+        # self._activation = pg_nn.GatedActivation()
+        self._kernel_size = kernel_size
+        self._padding = (kernel_size - 1) // 2  # (kernel_size - stride) / 2
+        self._mask_center = mask_center
+
+        # Vertical stack convolutions.
+        self._vstack_1xN = nn.Conv2d(
+            in_channels=self._in_channels,
+            out_channels=self._out_channels,
+            kernel_size=(1, self._kernel_size),
+            padding=(0, self._padding),
+        )
+        # TODO(eugenhotaj): Is it better to shift down the the vstack_Nx1 output
+        # instead of adding extra padding to the convolution? When we add extra
+        # padding, the cropped output rows will no longer line up with the rows of
+        # the vstack_1x1 output.
+        self._vstack_Nx1 = nn.Conv2d(
+            in_channels=self._out_channels,
+            out_channels=2 * self._out_channels,
+            kernel_size=(self._kernel_size // 2 + 1, 1),
+            padding=(self._padding + 1, 0),
+        )
+        self._vstack_1x1 = nn.Conv2d(
+            in_channels=in_channels, out_channels=2 * out_channels, kernel_size=1
+        )
+
+        self._link = nn.Conv2d(
+            in_channels=2 * out_channels, out_channels=2 * out_channels, kernel_size=1
+        )
+
+        # Horizontal stack convolutions.
+        self._hstack_1xN = nn.Conv2d(
+            in_channels=self._in_channels,
+            out_channels=2 * self._out_channels,
+            kernel_size=(1, self._kernel_size // 2 + 1),
+            padding=(0, self._padding + int(self._mask_center)),
+        )
+        self._hstack_residual = nn.Conv2d(
+            in_channels=out_channels, out_channels=out_channels, kernel_size=1
+        )
+        self._hstack_skip = nn.Conv2d(
+            in_channels=out_channels, out_channels=out_channels, kernel_size=1
+        )
+
+    def forward(self, vstack_input, hstack_input):
+        """Computes the forward pass.
+        Args:
+            vstack_input: The iin_channels=1,
+        out_channels=1,
+        n_gated=2,
+        gated_channels=16,
+        head_channels=64,
+        q_level=256,
+        """
+        _, _, h, w = vstack_input.shape  # Assuming NCHW.
+
+        # Compute vertical stack.
+        vstack = self._vstack_Nx1(self._vstack_1xN(vstack_input))[:, :, :h, :]
+        link = self._link(vstack)
+        vstack += self._vstack_1x1(vstack_input)
+        vstack = self._activation(vstack)
+
+        # Compute horizontal stack.
+        hstack = link + self._hstack_1xN(hstack_input)[:, :, :, :w]
+        hstack = self._activation(hstack)
+        skip = self._hstack_skip(hstack)
+        hstack = self._hstack_residual(hstack)
+        # NOTE(eugenhotaj): We cannot use a residual connection for causal layers
+        # otherwise we'll have access to future pixels.
+        if not self._mask_center:
+            hstack += hstack_input
+
+        return vstack, hstack, skip
+
+
+class GatedPixelCNN(nn.Module):
```

Comment:
Try to write unittest for `GatedPixelCNN`, you can refer to unittest example [here](https://github.com/opendilab/DI-engine/blob/main/ding/model/common/tests/test_encoder.py#L13)